File: apps/ios/Centwise/Core/Processor/SmsTransactionProcessor.swift
import Foundation

/** Thin iOS adapter: Shortcut/App Intent supplies the message; Rust parses and stores it. */
public final class SmsTransactionProcessor {
    public static let shared = SmsTransactionProcessor()

    public init() {}

    @discardableResult
    public func processIncomingSms(
        body: String,
        senderHint: String? = nil,
        date: Date = Date()
    ) -> SmsIngestResult? {
        let result = CentwiseRustBackend.ingestSMS(body: body, senderHint: senderHint, date: date)
        if let txId = result?.transactionId, let tx = CentwiseRustBackend.getTransaction(id: txId) {
            let isIncome = tx.kind == .income
            let catName = tx.categoryId.capitalized
            CentwiseNotifications.notifyIngestedTransaction(
                title: tx.title,
                amountMinor: tx.amountMinor,
                isIncome: isIncome,
                categoryName: catName,
                id: tx.id
            )
        }
        DispatchQueue.main.async {
            TransactionRepository.shared.loadFromRust()
            ReviewQueueRepository.shared.refresh()
        }
        return result
    }
}

File: apps/android/app/src/main/kotlin/com/centwise/core/processor/SmsTransactionProcessor.kt
package com.centwise.core.processor

import android.content.Context
import com.centwise.core.backend.CentwiseRustBackend
import com.centwise.core.uniffi.SmsIngestResult
import com.centwise.core.uniffi.SmsIngestStatus
import com.centwise.data.repository.TransactionRepository
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch

/** Thin Android adapter: capture details enter parser/Rust; persistence and UI update seamlessly. */
object SmsTransactionProcessor {
    fun processIncomingSms(
        context: Context,
        sender: String,
        body: String,
        timestamp: Long = System.currentTimeMillis(),
        notifyUser: Boolean = true
    ): SmsIngestResult? {
        CentwiseRustBackend.initialize(context.applicationContext)
        val result = CentwiseRustBackend.ingestSms(sender, body, timestamp)

        if (notifyUser && result != null && result.status == SmsIngestStatus.INSERTED && result.transactionId != null) {
            // Show transaction notification
            try {
                val record = CentwiseRustBackend.getTransaction(result.transactionId)
                if (record != null) {
                    com.centwise.core.notifications.CentwiseNotifications.notifyNewTransaction(
                        context = context.applicationContext,
                        transaction = com.centwise.data.models.TransactionItem(
                            id = record.id,
                            title = record.title,
                            amount = record.amountMinor / 100.0,
                            type = if (record.kind == com.centwise.core.uniffi.TransactionKind.INCOME) com.centwise.data.models.TransactionType.INCOME else com.centwise.data.models.TransactionType.EXPENSE,
                            category = record.categoryId,
                            paymentMethod = sender,
                            timestamp = record.occurredAtEpochMs,
                            note = record.notes,
                            rawSms = record.rawSms,
                            reference = record.reference
                        )
                    )
                }
            } catch (_: Exception) {}
        }

        // Notify and refresh UI StateFlows immediately
        if (notifyUser) {
            CoroutineScope(Dispatchers.Main).launch {
                TransactionRepository.shared.loadFromRust()
            }
        }

        return result
    }
}
